x-pack/plugin/esql/src/main/java/org/elasticsearch/xpack/esql/plugin/ComputeService.java

```diff
@@ -395,36 +401,121 @@ private void startComputeOnRemoteClusters(
                             TransportRequestOptions.EMPTY,
-                            new ActionListenerResponseHandler<>(clusterListener, ComputeResponse::new, esqlExecutor)
+                            new ActionListenerResponseHandler<>(
+                                listenerGroup.getClusterRequestListener(),
+                                ComputeResponse::new,
+                                esqlExecutor
+                            )
                         );
+                    }, e -> {
+                        // Exchange opening failed
+                        if (suppressRemoteFailure) {
+                            computeListener.markAsPartial(clusterAlias, e);
+                            openExchangeListener.onResponse(null);
+                        } else {
+                            openExchangeListener.onFailure(e);
+                        }
                     })
                 );
             }
         }
     }
 
+    // Create group task for this cluster. This group task ensures that two branches of the computation:
+    // the exchange sink and the cluster request, belong to the same group and each of them can cancel the other.
+    // runAfter listeners below ensure that the group is finalized when both branches are done.
+    // The group task is the child of the root task, so if the root task is cancelled, the group task is cancelled too.
+    private class RemoteListenerGroup {
+        private final CancellableTask groupTask;
+        private final ActionListener<Void> exchangeRequestListener;
+        private final ActionListener<ComputeResponse> clusterRequestListener;
+
+        public CancellableTask getGroupTask() {
+            return groupTask;
+        }
+
+        public ActionListener<Void> getExchangeRequestListener() {
+            return exchangeRequestListener;
+        }
+
+        public ActionListener<ComputeResponse> getClusterRequestListener() {
+            return clusterRequestListener;
+        }
+
+        RemoteListenerGroup(Task rootTask, ComputeListener computeListener, String clusterAlias, ActionListener<Void> delegate) {
+            final boolean suppressRemoteFailure = computeListener.shouldIgnoreRemoteErrors(clusterAlias);
+            groupTask = createGroupTask(rootTask, rootTask.getDescription() + "[" + clusterAlias + "]");
+            CountDown countDown = new CountDown(2);
+            // The group is done when both the sink and the cluster request are done
+            Runnable finishGroup = () -> {
+                if (countDown.countDown()) {
+                    taskManager.unregister(groupTask);
+                    delegate.onResponse(null);
+                }
+            };
+            // Cancel the group on sink failure
+            ActionListener<Void> exchangeListener = computeListener.acquireAvoid().delegateResponse((inner, e) -> {
+                taskManager.cancelTaskAndDescendants(groupTask, "exchange sink failure", true, ActionListener.noop());
```

Comment:
I think we should wait for the cancellation here. I think we should also fix the same issue in ComputerListener. Can you move this into the cancellation listener?

```java
if (suppressRemoteFailure) {
	computeListener.markAsPartial(clusterAlias, e);
	taskManager.cancelTaskAndDescendants(groupTask, "exchange sink failure", true, ActionListener.running(() -> inner.onResponse(null));
} else {
	inner.onFailure(e);
}
```

Reply:
Wait, wouldn't that code not cancel task on failure?